sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -231,7 +229,8 @@ class Dataset[T] private[sql](
       case _ =>
         queryExecution.analyzed
     }
-    if (sparkSession.sessionState.conf.getConf(SQLConf.FAIL_AMBIGUOUS_SELF_JOIN_ENABLED)) {
+    if (sparkSession.sessionState.conf.getConf(SQLConf.FAIL_AMBIGUOUS_SELF_JOIN_ENABLED) &&
+        plan.getTagValue(Dataset.DATASET_ID_TAG).isEmpty) {
```

Comment:
I find that `df.show()` can also change the dataset_id of LogicalPlan. So I made this change.

Reply:
Any test fails? If not, could you add tests for this change?

Reply:
Added, thanks!